src/io_family.f90

```diff
@@ -0,0 +1,586 @@
+!! SPDX-License-Identifier: BSD-3-Clause
+
+!
+! This is the IO module for families of readers / writers
+!    => Initialize / finalize a family of readers / writers
+!    => Register a variable in a family of readers / writers
+!
+
+module decomp_2d_io_family
+
+   use decomp_2d_constants
+   use decomp_2d_mpi
+   use decomp_2d_io_utilities
+   use decomp_2d
+   use MPI
+#ifdef ADIOS2
+   use adios2
+#endif
+   use, intrinsic :: iso_fortran_env, only: real32, real64
+
+   implicit none
+
+#ifdef ADIOS2
+   type(adios2_adios), save, pointer :: default_adios
+#endif
+
+   !
+   ! derived type to store info for a family of readers / writers
+   !
+   type, public :: d2d_io_family
+      integer :: type = decomp_2d_io_none                     ! Type of the writer
+      character(:), allocatable :: label                      ! Label of the writer
+#ifdef ADIOS2
+      type(adios2_adios), pointer :: adios                    ! adios2 only
+      type(adios2_io) :: io                                   ! adios2 only
+#endif
+   contains
+      procedure :: init => d2d_io_family_init                     ! ADIOS2 writer if possible, MPI otherwise
+      procedure :: mpi_init => d2d_io_family_mpi_init             ! Force MPI writer
+      procedure :: adios2_init => d2d_io_family_adios2_init       ! Force ADIOS2 writer
+      procedure :: fin => d2d_io_family_fin                       ! Clear the writer
+      procedure :: register_var3d => d2d_io_family_register_var3d ! Register a 3D variable
+      procedure :: register_var2d => d2d_io_family_register_var2d ! Register a 2D variable
+      procedure :: log => d2d_io_family_log                       ! Print some information about the object
+   end type d2d_io_family
+
+   private
+
+   public :: decomp_2d_io_family_init, &
+             decomp_2d_io_family_fin
+#ifdef ADIOS2
+   public :: decomp_2d_io_family_get_default_adios, &
+             decomp_2d_io_family_set_default_adios
+#endif
+
+contains
+
+   !
+   ! Initialize a new family of readers / writers (default type)
+   !
+   subroutine d2d_io_family_init(family, label)
```

Comment:
Could this receive an (optional) argument to request a specific backend? e.g.
```
call d2d_io_family_init(family, label, backend=D2D_ADIOS2_BACKEND)
```

Reply:
The current design allow the external code to select the backend using `family%mpi_init` or `family%adios_init`. The addition of one optional parameter for calls to `family%init` is a simple addition and I will implement it. The allowed values for the optional argument will be `DECOMP_2D_IO_MPI` and `DECOMP_2D_IO_ADIOS2`.

Reply:
Fixed in https://github.com/2decomp-fft/2decomp-fft/pull/269/commits/482a23bdc19b8877906f901558c2cd466c921b36